File: Java/src/session3/td/UnionIter.java
package session3.td;

public class UnionIter implements Mot{
	private Mot gauche;
	private Mot droit;

	public UnionIter(Mot g, Mot d) {
		this.gauche = g;
		this.droit = d;
	}

	@Override
	public int taille() {
		int result = 0;
		Mot tempGauche = gauche;
		while(!tempGauche.casVide()) {
			result++;
			tempGauche = tempGauche.reste();
		}
		Mot tempDroit = droit;
		while(!tempDroit.casVide()) {
			result++;
			tempDroit = tempDroit.reste();
		}
		return result;
	}

	@Override
	public boolean casUnion() {
		return true;
	}

	@Override
	public char caractere() {
		Iterateur i = this.iterateur();
		return i.suivant();
	}

	@Override
	public Mot reste() {
		Iterateur i = this.iterateur();
		return i.reste();
	}

	@Override
	public Mot gauche() {
		return this.gauche;
	}

	@Override
	public Mot droit() {
		return this.droit;
	}

	
}

File: TD3/typescript/session3/td/UnionIter.ts
import { Mot } from "./Mot";
import { Iterateur } from "./Iterateur";

export class UnionIter extends Mot {
    private motGauche: Mot;
	private motDroit: Mot;
    private vTaille: number;

    public constructor(g:Mot, d:Mot){
        super();
        this.motGauche=g;
        this.motDroit=d;
        this.vTaille = this.motGauche.taille() + this.motDroit.taille();
    }


    taille(): number {
    //     var result = 0;
	// 	var tempGauche: Mot = this.motGauche;
	// 	while(!tempGauche.casVide()) {
	// 		result++;
	// 		tempGauche = tempGauche.reste();
	// 	}
	// 	var tempDroit = this.motDroit;
	// 	while(!tempDroit.casVide()) {
	// 		result++;
	// 		tempDroit = tempDroit.reste();
	// 	}
    // return result;
    return this.vTaille;
    }

    casUnion():boolean{
        return true;
    }

    reste():Mot{
        console.log("Passe par reste");
        var i:Iterateur = this.iterateur();
        return i.reste();
    }

    caractere():string{
        var i:Iterateur = this.iterateur();
        return i.next().value;
    }

    gauche():Mot{
        return this.motGauche;
    }

    droit():Mot{
        return this.motDroit;
    }
}